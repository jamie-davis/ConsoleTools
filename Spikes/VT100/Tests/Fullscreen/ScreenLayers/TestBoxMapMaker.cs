using TestConsoleLib.Testing;
using VT100.FullScreen.ScreenLayers;
using VT100.Tests.Utilities;
using Xunit;

namespace VT100.Tests.Fullscreen.ScreenLayers
{
    public class TestBoxMapMaker
    {
        [Fact]
        public void BoxRegionIsMapped()
        {
            //Arrange
            var region = new BoxRegion(1, 1, 5, 5);
            
            //Act
            var map = BoxMapMaker.Map(new[] { region }, 10, 10);

            //Assert
            var plate = new Plate(10, 10);
            BoxRenderer.RenderMapToPlate(map, plate);
            PlateDumpFormatter.Format(plate).Verify();
        }

        [Fact]
        public void BoxRegionThatExceedsRightEdgeIsTrimmed()
        {
            //Arrange
            var region = new BoxRegion(7, 1, 5, 5);
            
            //Act
            var map = BoxMapMaker.Map(new[] { region }, 10, 10);

            //Assert
            var plate = new Plate(10, 10);
            BoxRenderer.RenderMapToPlate(map, plate);
            PlateDumpFormatter.Format(plate).Verify();
        }

        [Fact]
        public void BoxRegionThatExceedsBottomEdgeIsTrimmed()
        {
            //Arrange
            var region = new BoxRegion(1, 7, 5, 5);
            
            //Act
            var map = BoxMapMaker.Map(new[] { region }, 10, 10);

            //Assert
            var plate = new Plate(10, 10);
            BoxRenderer.RenderMapToPlate(map, plate);
            PlateDumpFormatter.Format(plate).Verify();
        }

        [Fact]
        public void BoxRegionThatExceedsTopEdgeIsTrimmed()
        {
            //Arrange
            var region = new BoxRegion(1, -2, 5, 5);
            
            //Act
            var map = BoxMapMaker.Map(new[] { region }, 10, 10);

            //Assert
            var plate = new Plate(10, 10);
            BoxRenderer.RenderMapToPlate(map, plate);
            PlateDumpFormatter.Format(plate).Verify();
        }

        [Fact]
        public void BoxRegionOnlyNeedsOneCharacterInView()
        {
            //Arrange
            var regions = new []
            {
                new BoxRegion(-4, -4, 5, 5),
                new BoxRegion(9, 9, 5, 5),
                new BoxRegion(9, -4, 5, 5),
                new BoxRegion(-4, 9, 5, 5)
            };
            
            //Act
            var map = BoxMapMaker.Map(regions, 10, 10);

            //Assert
            var plate = new Plate(10, 10);
            BoxRenderer.RenderMapToPlate(map, plate);
            PlateDumpFormatter.Format(plate).Verify();
        }

        [Fact]
        public void BoxRegionThatIsOffScreenToBottomAndRightIsSkipped()
        {
            //Arrange
            var region = new BoxRegion(11, 11, 5, 5);
            
            //Act
            var map = BoxMapMaker.Map(new[] { region }, 10, 10);

            //Assert
            var plate = new Plate(10, 10);
            BoxRenderer.RenderMapToPlate(map, plate);
            PlateDumpFormatter.Format(plate).Verify();
        }

        [Fact]
        public void BoxRegionsEdgesCanOverlap()
        {
            //Arrange
            var regions = new []
            {
                new BoxRegion(0, 0, 3, 3),
                new BoxRegion(2, 0, 3, 3),
                new BoxRegion(4, 0, 3, 3),
                new BoxRegion(6, 0, 3, 3),
                new BoxRegion(8, 0, 3, 3),
                new BoxRegion(1, 2, 3, 3),
                new BoxRegion(3, 2, 3, 3),
                new BoxRegion(5, 2, 3, 3),
                new BoxRegion(7, 2, 3, 3),
                new BoxRegion(1, 4, 3, 3),
                new BoxRegion(3, 4, 3, 3),
                new BoxRegion(5, 4, 3, 3),
                new BoxRegion(7, 4, 3, 3),
            };
            
            //Act
            var map = BoxMapMaker.Map(regions, 10, 10);

            //Assert
            var plate = new Plate(10, 10);
            BoxRenderer.RenderMapToPlate(map, plate);
            PlateDumpFormatter.Format(plate).Verify();
        }
    }
}