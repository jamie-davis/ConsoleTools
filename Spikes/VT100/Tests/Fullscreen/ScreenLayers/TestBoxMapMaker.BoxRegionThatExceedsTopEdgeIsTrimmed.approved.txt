Plate dimensions: x10 y10

    Text           Colour    
 +0123456789+    +0123456789+
0| │   │    |   0|          |
1| │   │    |   1|          |
2| └───┘    |   2|          |
3|          |   3|          |
4|          |   4|          |
5|          |   5|          |
6|          |   6|          |
7|          |   7|          |
8|          |   8|          |
9|          |   9|          |
 +0123456789+    +0123456789+
